docs/proposals/scheduling/20220701-pod-migration-job.md

```diff
@@ -0,0 +1,354 @@
+
+Therefore, it is necessary to realize a final state-oriented migration capability based on CRD, track the status of each process in the migration, and perceive scenarios such as upgrading and scaling of the application.
+
+### Goals
+
+1. Defines a CRD-based Pod Migration Job API, through which the descheduler can evict or delete Pods more safely.
+2. Describe in detail the design details behind the API.
+
+### Non-Goals/Future Work
+
+1. A new descheduler framework
+2. Descheduling capability for different scenarios such as load-aware descheduling, defragemention, etc.
+3. The details about Deterministic preemption that preempts other Pods for Reservation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+The descheduler in the K8s community evicts pods to be rescheduled according to different strategies. However, it does not guarantee whether the evicted Pod has resources available after re-creation. If a large number of new Pods are in the Pending state when the resources in the cluster are tight, it may lead to application availability.
+
+#### Story 2
+
+The descheduler evicts the Pod through the Eviction API, and the Eviction API decides whether to delete the Pod according to the PDB status. However, it is unable to perceive workload upgrades, scaling and other scenarios in which Pods are deleted, which will also bring security risks.
+
+#### Story 3
+
+The Pod migration capability itself can be provided to users as a service. Users can integrate this API in their own systems to achieve safe migration, and are no longer limited to deschedulers.
+
+
+### Basic Migration API
+
+These APIs provide cluster administrators with more fine-grained migration control capabilities, which can better reduce risks.
+
+- Label `scheduling.koordinator.sh/refuse-to-migrate` indicates that it refuses to migrate this Pod. Although K8s assumes that all Pods should be migrated/evicted, there are always some practical exceptions, such as some stateful applications that may have serious problems if migrated, or some applications that are not yet capable of being migrated .
+- `scheduling.koordinator.sh/migration-priority` indicates the order of migration. Although the K8s community has [Pod Deletion Cost #2255](https://github.com/kubernetes/enhancements/issues/2255), it is not a general mechanism. To avoid conflicts with components that use `Pod Deletion Cost`, users can individually mark migration priorities for Pods. The higher the value, the later in the migration order.
```

Comment:
These two fields may be merged, and the cost is set to max when migration is not allowed.

scheduling.koordinator.sh/eviction-cost

When we have other SLO actions, we can trigger Pod eviction, and call it eviction-cost directly?

Reply:
For example, when the node interference is serious, and we need to evict some Pods, we hope that this value can be reused

Reply:
Total agree.